### ISSUE TITLE
[Question] How can I use component testing with Redux components

### ISSUE THREAD

**Issue by hippyhorse** [microsoft/playwright#15325]
## Component Testing with Redux

Super excited to see [component testing](https://www.youtube.com/watch?v=y3YxX4sFJbM) in preview! This is working great for some of our simple components, but we are struggling with how to use it for components that use Redux.

## Reproduction steps

Go to an empty folder and do the following...

1. Initialise Playwright with component testing:

```sh
npm init playwright@latest -- --ct
```

2. Install dependencies

```sh
npm install react
npm install react-dom
npm install react-redux
npm install @reduxjs/toolkit
```

3. Create three the files below: `Simple.tsx`, `Simple.spec.tsx` and `counterSlice.ts`.

4. Run the test

You should see the following (unexpected) error:

```
undefined: TypeError: Cannot read properties of undefined (reading 'counter')
    at selectCounterValue (http://localhost:3100/assets/index.b9c26efb.js:4745:45)
    at a (http://localhost:3100/assets/index.b9c26efb.js:569:270)
    at http://localhost:3100/assets/index.b9c26efb.js:569:489
    at Object.useSyncExternalStore (http://localhost:3100/assets/index.b9c26efb.js:346:377)
    at react_production_min.useSyncExternalStore (http://localhost:3100/assets/index.b9c26efb.js:107:709)
    ...
```

I've been hitting my head against a wall on this, and can't work out why it won't work. The origin of this is that I want to move away from Jest for my component testing; but that means I need to get Redux components working.

### Component (`Simple.tsx`)

```ts
import { useDispatch, useSelector } from 'react-redux';
import { increment, decrement, selectCounterValue } from './counterSlice';

export default function Simple() {
    const dispatch = useDispatch();
    const value = useSelector(selectCounterValue);

    return (
        <div>
            <p>Counter: {value}</p>
            <button id="inc" onClick={() => dispatch(increment())}>+</button>
            <button id="dec" onClick={() => dispatch(decrement())}>-</button>
        </div >
    );
}
```

### Redux slice (`counterSlice.ts`)

```ts
import { configureStore, createSlice } from '@reduxjs/toolkit';

export const counterSlice = createSlice({
    name: 'counter',
    initialState: {
        value: 0
    },
    reducers: {
        increment(state) {
            state.value++
        },
        decrement(state) {
            state.value--
        },
    }
});

export const store = configureStore({
    reducer: {
        counter: counterSlice.reducer
    }
});

export const { increment, decrement } = counterSlice.actions;
export const selectCounterValue = (state) => state.counter.value;
```

### Test (`Simple.spec.tsx`)

```ts
import { expect, test } from '@playwright/experimental-ct-react';
import { Provider } from 'react-redux';
import { store } from './counterSlice';

import Simple from './Simple';

test('Simple test', async ({ mount }) => {
    const el = await mount(
        <Provider store={store} >
            <Simple />
        </Provider>
    );

    await expect(el).toContainText('0');
    await el.locator('#inc').click();
    await expect(el).toContainText('1');
    await el.locator('#dec').click();
    await expect(el).toContainText('0');
});
```

**Comment by rwoll** [comment]
Whatever components and properties appear directly in your `*.spec.tsx`, Playwright serializes parts of them between the Node test runner and the Browser. Since `store` is a complex object that likely only wants to live in the browser, something is breaking under the hood. See some notes here https://playwright.dev/docs/test-components#under-the-hood as well as the _Under the Hood: How does some of the new magic work? Looking at the mount fixture under the hood (with slides!)_ section in https://www.youtube.com/watch?v=y3YxX4sFJbM&t=850s for more information. Playwright can serialize: JSX Components and JSON serializable properties (along with functions).

<img width="1209" alt="Screen Shot 2022-07-02 at 3 00 43 PM" src="https://user-images.githubusercontent.com/11915034/177017284-77357079-adea-489d-aefd-ffb6df124940.png">
<img width="1231" alt="Screen Shot 2022-07-02 at 3 00 59 PM" src="https://user-images.githubusercontent.com/11915034/177017287-31c5be12-a775-4c53-93f9-3a4d7eaf8b93.png">


For now, here's a **workaround**[^1]:

Change `Simple.spec.tsx` to:

```tsx
import { expect, test } from '@playwright/experimental-ct-react';
import { WrappedComponent } from './Wrapper';

test('Simple test', async ({ mount }) => {
    const el = await mount(<WrappedComponent />);

    await expect(el).toContainText('0');
    await el.locator('#inc').click();
    await expect(el).toContainText('1');
    await el.locator('#dec').click();
    await expect(el).toContainText('0');
});
```

and `Wrapper.tsx` would be:

```ts
import { Provider } from "react-redux";
import Simple from "./Simple";
import { store } from './counterSlice';

export const WrappedComponent = () => {
  return (
    <Provider store={store} ><Simple /></Provider>
   );
}
```

In other words, keep store out of `store` the `*.spec.tsx` file.

The test then passes. Unfortunately, it doesn't give your direct access to your store in Node.

[^1]: Let me know if this works for you!

**Comment by rwoll** [comment]
1.25 label added since—while we can't fully "fix" this—we can better document it.

**Comment by hippyhorse** [comment]
@rwoll Thanks! Yes this is working well now.

I'm not onto another hurdle (`somefile.ts:5 Uncaught ReferenceError: Cannot access 'somemodule' before initialization`)

Looks like it's a possible circular dependency in my code that is causes issues with Vite (https://github.com/vitejs/vite/issues/3033#issuecomment-973247243) - I'll update this thread once confirmed one way or another.

**Comment by pavelfeldman** [comment]
Closing as per above, let's embrace the wrapper solution. Please let us know if there are outstanding issues.
